src/app/services/actions/query-action-creators.ts

```diff
@@ -1,28 +1,80 @@
+import { writeData } from '../../../store/cache';
+import { IHistoryItem } from '../../../types/history';
 import { IQuery } from '../../../types/query-runner';
+import { createHarPayload, generateHar } from '../../views/sidebar/history/historyUtil';
 import { queryResponseError } from './error-action-creator';
-import { anonymousRequest, authenticatedRequest, parseResponse, queryResponse } from './query-action-creator-util';
+import {
+  anonymousRequest, authenticatedRequest, isImageResponse,
+  parseResponse, queryResponse
+} from './query-action-creator-util';
+import { addHistoryItem } from './request-history-action-creators';
 
 export function runQuery(query: IQuery): Function {
   return (dispatch: Function, getState: Function) => {
     const tokenPresent = getState().authToken;
     const respHeaders: any = {};
+    const createdAt = new Date().toISOString();
 
     if (tokenPresent) {
       return authenticatedRequest(dispatch, query).then(async (response: Response) => {
-
-        if (response && response.ok) {
-          const result = await parseResponse(response, respHeaders);
-          return dispatch(
-            queryResponse({
-              body: result,
-              headers: respHeaders
-            }),
-          );
-        }
-        return dispatch(queryResponseError(response));
+        await processResponse(response, respHeaders, dispatch, createdAt);
       });
     }
 
-    return anonymousRequest(dispatch, query);
+    return anonymousRequest(dispatch, query).then(async (response: Response) => {
+      await processResponse(response, respHeaders, dispatch, createdAt);
+    });
+  };
+
+  async function processResponse(response: Response, respHeaders: any, dispatch: Function, createdAt: any) {
+
+    const result = await parseResponse(response, respHeaders);
+    createHistory(response, respHeaders, query, createdAt, dispatch, result);
+
+    if (response && response.ok) {
+      return dispatch(queryResponse({
+        body: result,
+        headers: respHeaders
+      }));
+    }
+    else {
+      return dispatch(queryResponseError(response));
+    }
+
+  }
+}
+
+async function createHistory(response: Response, respHeaders: any, query: IQuery,
+  createdAt: any, dispatch: Function, result: any) {
+  const status = response.status;
+  const statusText = response.statusText;
+  const { duration, contentType } = respHeaders;
+
+  if (isImageResponse(contentType)) {
+    result = await result.clone().arrayBuffer();
+  }
+
+  const historyItem: IHistoryItem = {
+    url: query.sampleUrl,
+    method: query.selectedVerb,
+    headers: query.sampleHeaders,
+    body: query.sampleBody,
+    responseHeaders: respHeaders,
+    createdAt,
+    status,
+    statusText,
+    result,
+    duration,
+    har: ''
   };
+
```

Comment:
Should we create the HAR payload if we are not exporting/downloading history items?

Reply:
The HAR payload was created but the export function did not work. It was removed from the UI to make it work as another work item.

Reply:
> I don't understand why click on view in the history component dispatches QUERY_GRAPH_SUCCESS and it's not obvious from the code. Please respond on how/where this is happening.

> 

View dispatches an action creator called viewHistoryItem which dispatches the QUERY_GRAPH_SUCCESS action.

Reply:
I see, this is misleading and has an element of surprise to it. The action type `QUERY_GRAPH_SUCCESS` should only be dispatched when making a network request, which is not the case for viewing history items.

I suggest that you dispatch a `VIEW_HISTORY_ITEM_SUCCESS` action type and handle the type in the graphResponse reducer.